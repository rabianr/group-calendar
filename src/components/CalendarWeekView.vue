<script setup lang="ts">
import { computed } from 'vue'
import {
  isSunday,
  isSaturday,
  isBefore,
  isAfter,
  addDays,
  differenceInDays,
  min as minDate,
  format,
  parseISO
} from 'date-fns'
import { useI18n } from 'vue-i18n'
import { useCalendarStore } from '@/stores/calendar'
import type { GCEvent, DisplayEvent } from './CalendarEvent.types'
import CalendarEvent from './CalendarEvent.vue'
import I18nDateDisplay from './I18nDateDisplay.vue'
import IconTrash3 from './IconTrash3.vue'

const props = defineProps<{
  start: Date
  end: Date
}>()

const { t } = useI18n()
const calendar = useCalendarStore()

// start から end までの日付配列を生成
const dates = computed(() => {
  const arr = []

  for (let curr = props.start; !isAfter(curr, props.end); curr = addDays(curr, 1)) {
    arr.push(curr)
  }

  return arr
})

// 表示のため、データのグルーピング、イベントの配置位置を事前に計算
const memberData = computed(() => {
  // 返却値
  const ret: { [key: string]: { [key: string]: DisplayEvent[] } } = {}
  // 一つのセルに配置したイベントの配列、値はセルの連結値（colspan）
  const currPlacements: number[] = []

  /**
   * イベントが配置可能な位置を検索
   *
   * @param colspan 横方向のセルの連結
   * @returns 0ベースのインデックス位置
   */
  const findPlaceablePos = (colspan: number): number => {
    let pos = 0

    while (currPlacements[pos]) {
      pos++
    }

    currPlacements[pos] = colspan

    return pos
  }

  /**
   * 現セルに全イベントが配置完了したら、各 colspan を１引いて、次のセル配置に備える
   */
  const finishPlacing = () => {
    for (let key in currPlacements) {
      if (currPlacements[key] > 0) {
        currPlacements[key]--
      }
    }
  }

  for (let member in calendar.members) {
    const groupedEvents: { [key: string]: (GCEvent & { colspan: number })[] } = {}

    // イベントを開始日毎にグルーピング、セルの連結を計算
    for (let event of calendar.members[member]) {
      const evStart = parseISO(event.start)
      const evEnd = parseISO(event.end)

      // イベント開始日は表示範囲内かチェック
      if (!isBefore(evStart, props.start) && !isAfter(evStart, props.end)) {
        if (!groupedEvents[event.start]) {
          groupedEvents[event.start] = []
        }

        groupedEvents[event.start].push({
          ...event,
          colspan: Math.min(7, differenceInDays(minDate([evEnd, props.end]), evStart) + 1)
        })

        // イベント開始日は表示開始日以前とイベント終了日は表示開始日以降かチェック
      } else if (isBefore(evStart, props.start) && !isBefore(evEnd, props.start)) {
        const dKey = format(props.start, 'yyyy-MM-dd')

        if (!groupedEvents[dKey]) {
          groupedEvents[dKey] = []
        }

        groupedEvents[dKey].push({
          ...event,
          colspan: Math.min(7, differenceInDays(evEnd, props.start) + 1)
        })
      }
    }

    ret[member] = {}

    // セル毎にイベントの配置位置を計算
    for (let d of dates.value) {
      const dKey = format(d, 'yyyy-MM-dd')
      const arr: DisplayEvent[] = []

      if (groupedEvents[dKey]) {
        // 降順でセルの連結で並び替え
        groupedEvents[dKey].sort((a, b) => {
          if (a.colspan == b.colspan) {
            return a.start > b.start ? 1 : -1
          }

          return a.colspan > b.colspan ? -1 : 1
        })

        for (let i = 0; i < groupedEvents[dKey].length; i++) {
          arr.push({
            ...groupedEvents[dKey][i],
            pos: findPlaceablePos(groupedEvents[dKey][i].colspan),
            last: i == groupedEvents[dKey].length - 1
          })
        }
      }

      finishPlacing()

      ret[member][dKey] = arr
    }
  }

  return ret
})

const deleteMember = (name: string) => {
  const confirmed = confirm(t('Are you sure you want to delete this member?'))

  if (confirmed) {
    calendar.deleteMember(name)
  }
}
</script>

<template>
  <table class="table table-bordered border-secondary-subtle">
    <thead>
      <tr>
        <th class="col-header text-center">{{ $t('Member') }}</th>
        <th
          v-for="d in dates"
          :key="d.getTime()"
          class="col-header text-center"
          :class="{ 'text-danger': isSunday(d), 'text-info': isSaturday(d) }"
        >
          <I18nDateDisplay :date="d" format-str="M/d (E)" />
        </th>
      </tr>
    </thead>
    <tbody>
      <tr v-for="(eventData, member) in memberData" :key="member">
        <td>
          {{ member }}
          <div class="mt-1">
            <button
              type="button"
              class="btn btn-link text-danger p-0"
              @click="() => deleteMember(member as string)"
            >
              <IconTrash3 />
            </button>
          </div>
        </td>
        <td class="col-events" v-for="(events, date) in eventData" :key="date">
          <CalendarEvent v-for="(event, index) in events" :key="index" :data="event" />
        </td>
      </tr>
    </tbody>
  </table>
</template>

<style lang="scss" scoped>
.table {
  table-layout: fixed;

  .col-header {
    font-weight: var(--bs-body-font-weight);
  }

  .col-events {
    --evt-height: 23px;
    --evt-margin: 2px;
    position: relative;
    height: 90px;
    padding: 0;
  }
}
</style>
