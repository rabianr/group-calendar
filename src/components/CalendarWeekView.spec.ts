import { shallowMount } from '@vue/test-utils'
import { describe, expect, it } from 'vitest'
import { createPinia, setActivePinia } from 'pinia'
import CalendarWeekView from './CalendarWeekView.vue'

describe('CalendarWeekView コンポーネント', () => {
  setActivePinia(createPinia())

  it('スナップショットテスト', () => {
    const wrapper = shallowMount(CalendarWeekView, {
      props: {
        start: new Date(2024, 3, 10),
        end: new Date(2024, 3, 16)
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})
