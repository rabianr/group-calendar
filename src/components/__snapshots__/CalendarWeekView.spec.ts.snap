// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CalendarWeekView コンポーネント > スナップショットテスト 1`] = `
"<table data-v-39b37b02="" class="table table-bordered border-secondary-subtle">
  <thead data-v-39b37b02="">
    <tr data-v-39b37b02="">
      <th data-v-39b37b02="" class="col-header text-center">メンバー</th>
      <th data-v-39b37b02="" class="col-header text-center">
        <i18n-date-display-stub data-v-39b37b02="" date="Wed Apr 10 2024 00:00:00 GMT+0900 (Japan Standard Time)" formatstr="M/d (E)"></i18n-date-display-stub>
      </th>
      <th data-v-39b37b02="" class="col-header text-center">
        <i18n-date-display-stub data-v-39b37b02="" date="Thu Apr 11 2024 00:00:00 GMT+0900 (Japan Standard Time)" formatstr="M/d (E)"></i18n-date-display-stub>
      </th>
      <th data-v-39b37b02="" class="col-header text-center">
        <i18n-date-display-stub data-v-39b37b02="" date="Fri Apr 12 2024 00:00:00 GMT+0900 (Japan Standard Time)" formatstr="M/d (E)"></i18n-date-display-stub>
      </th>
      <th data-v-39b37b02="" class="col-header text-center text-info">
        <i18n-date-display-stub data-v-39b37b02="" date="Sat Apr 13 2024 00:00:00 GMT+0900 (Japan Standard Time)" formatstr="M/d (E)"></i18n-date-display-stub>
      </th>
      <th data-v-39b37b02="" class="col-header text-center text-danger">
        <i18n-date-display-stub data-v-39b37b02="" date="Sun Apr 14 2024 00:00:00 GMT+0900 (Japan Standard Time)" formatstr="M/d (E)"></i18n-date-display-stub>
      </th>
      <th data-v-39b37b02="" class="col-header text-center">
        <i18n-date-display-stub data-v-39b37b02="" date="Mon Apr 15 2024 00:00:00 GMT+0900 (Japan Standard Time)" formatstr="M/d (E)"></i18n-date-display-stub>
      </th>
      <th data-v-39b37b02="" class="col-header text-center">
        <i18n-date-display-stub data-v-39b37b02="" date="Tue Apr 16 2024 00:00:00 GMT+0900 (Japan Standard Time)" formatstr="M/d (E)"></i18n-date-display-stub>
      </th>
    </tr>
  </thead>
  <tbody data-v-39b37b02="">
    <tr data-v-39b37b02="">
      <td data-v-39b37b02="">山田花子 <div data-v-39b37b02="" class="mt-1"><button data-v-39b37b02="" type="button" class="btn btn-link text-danger p-0">
            <icon-trash3-stub data-v-39b37b02=""></icon-trash3-stub>
          </button></div>
      </td>
      <td data-v-39b37b02="" class="col-events"></td>
      <td data-v-39b37b02="" class="col-events"></td>
      <td data-v-39b37b02="" class="col-events"></td>
      <td data-v-39b37b02="" class="col-events"></td>
      <td data-v-39b37b02="" class="col-events"></td>
      <td data-v-39b37b02="" class="col-events"></td>
      <td data-v-39b37b02="" class="col-events"></td>
    </tr>
    <tr data-v-39b37b02="">
      <td data-v-39b37b02="">山田太郎 <div data-v-39b37b02="" class="mt-1"><button data-v-39b37b02="" type="button" class="btn btn-link text-danger p-0">
            <icon-trash3-stub data-v-39b37b02=""></icon-trash3-stub>
          </button></div>
      </td>
      <td data-v-39b37b02="" class="col-events"></td>
      <td data-v-39b37b02="" class="col-events"></td>
      <td data-v-39b37b02="" class="col-events"></td>
      <td data-v-39b37b02="" class="col-events"></td>
      <td data-v-39b37b02="" class="col-events"></td>
      <td data-v-39b37b02="" class="col-events"></td>
      <td data-v-39b37b02="" class="col-events"></td>
    </tr>
    <tr data-v-39b37b02="">
      <td data-v-39b37b02="">何野某 <div data-v-39b37b02="" class="mt-1"><button data-v-39b37b02="" type="button" class="btn btn-link text-danger p-0">
            <icon-trash3-stub data-v-39b37b02=""></icon-trash3-stub>
          </button></div>
      </td>
      <td data-v-39b37b02="" class="col-events"></td>
      <td data-v-39b37b02="" class="col-events"></td>
      <td data-v-39b37b02="" class="col-events"></td>
      <td data-v-39b37b02="" class="col-events"></td>
      <td data-v-39b37b02="" class="col-events"></td>
      <td data-v-39b37b02="" class="col-events"></td>
      <td data-v-39b37b02="" class="col-events"></td>
    </tr>
    <tr data-v-39b37b02="">
      <td data-v-39b37b02="">権兵衛 <div data-v-39b37b02="" class="mt-1"><button data-v-39b37b02="" type="button" class="btn btn-link text-danger p-0">
            <icon-trash3-stub data-v-39b37b02=""></icon-trash3-stub>
          </button></div>
      </td>
      <td data-v-39b37b02="" class="col-events"></td>
      <td data-v-39b37b02="" class="col-events"></td>
      <td data-v-39b37b02="" class="col-events"></td>
      <td data-v-39b37b02="" class="col-events"></td>
      <td data-v-39b37b02="" class="col-events"></td>
      <td data-v-39b37b02="" class="col-events"></td>
      <td data-v-39b37b02="" class="col-events"></td>
    </tr>
  </tbody>
</table>"
`;
